c++/src/kj/buffering.h

```diff
@@ -0,0 +1,202 @@
+  OneOf<FdRangeInputBuffer*, Own<FdRangeInputBuffer>, Empty> parentOrOverflow;
+
+  FdRangeInputBuffer& getBuffer();
+
+  struct ReadParams {
+    byte* start;
+    size_t min;
+    size_t max;
+    ArrayPtr<AutoCloseFd> fds;
+  };
+  ReadParams getReadLocation(size_t min);
+
+  struct ReadResult {
+    size_t byteCount;
+    size_t capCount;
+  };
+  void handleReadResult(ReadResult result);
+};
+
+class EmptyFdBufferRange final {
+  // A handle that allows reading into a `FdRangeInputBuffer`.
+
+public:
+  inline Promise<void> readFromStream(
+      OneOf<AsyncInputStream*, AsyncIoStream*, AsyncCapabilityStream*> stream) {
+    KJ_SWITCH_ONEOF(stream) {
+      KJ_CASE_ONEOF(inputStream, AsyncInputStream*) {
+        return readFromStream(*inputStream);
+      }
+      KJ_CASE_ONEOF(ioStream, AsyncIoStream*) {
+        return readFromStream(*ioStream);
+      }
+      KJ_CASE_ONEOF(capStream, AsyncCapabilityStream*) {
+        return readFromStream(*capStream);
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+
+  Promise<void> readFromStream(AsyncInputStream& stream);
+  Promise<void> readFromStream(AsyncCapabilityStream& stream);
+  // Read as many bytes as possible from the provided stream into the buffer
+
+private:
+  EmptyFdBufferRange(FdRangeInputBuffer& parent);
+  friend class FdRangeInputBuffer;
+
+  FdRangeInputBuffer& parent;
+};
+
+class FdRangeInputBuffer {
+  // A buffer of up to two FD ranges that can be filled by reading from a stream.
+  // An FD range is defined as a range of bytes with FDs associated with the last byte in the range.
+  // FDs being associated with the last byte of the range is for the purpose of the buffering logic,
+  // resulting from the rules of SCM_RIGHTS control messages where a read that includes FDs will not
+  // read past the last byte written by the other side of the socket. One can infer from this rule,
+  // that if the other side of the socket sent multiple messages, any FDs received are part of the
+  // last message.
+  //
+  // The buffer is emptied by splitting off prefixes of it. If a split-off prefix includes the last
+  // byte of an FD range with FDs, those FDs are also split off.
+  //
+  // The limit of two buffered FD ranges results from the optimization described in `appendFromStream`.
+  // It's necessary to create a new FD range if the existing range already had FDs in it,
+  // otherwise it's safe to extend the existing range.
+  //
+  // It is assumed that when two FD ranges are buffered, once FDs are associated with the second range,
+  // then a complete "message" is present in some prefix of the buffer, and this prefix is >= the
+  // size of the first FD range. The caller must split this prefix out of the buffer
+  // before attempting to read any more bytes into the buffer.
+
+public:
+  KJ_DISALLOW_COPY(FdRangeInputBuffer);
+  FdRangeInputBuffer(size_t size, size_t maxFds);
+
+  OneOf<EmptyFdBufferRange, Own<SplittableFdBufferRange>> getRange();
```

Comment:
I'd like us to reconceptualize this as a generator class. Essentially, we should be able to say `getNextMessageRange()` without making the caller deal with a `OneOf`. We should be able to do the split for callers and not have a tightly coupled relationship with downstream code. Put another way, this doesn't have to be isomorphic to the input end of `AsyncIoMessageStream` but I do think it is an async input stream of some sort.